v0/destinations/singular/transform.js

```diff
@@ -0,0 +1,215 @@
+        SINGULAR_SESSION_ANDROID_EXCLUSION
+      );
+    } else if (platform.toLowerCase() === "ios") {
+      payload = constructPayload(
+        message,
+        MAPPING_CONFIG[CONFIG_CATEGORIES.SESSION_IOS.name]
+      );
+      if (!payload) {
+        // fail-safety for developer error
+        throw new CustomError("Failed to Create iOS Session Payload", 400);
+      }
+      extractCustomFields(
+        message,
+        eventAttributes,
+        ["properties"],
+        SINGULAR_SESSION_IOS_EXCLUSION
+      );
+
+      /*
+        if att_authorization_status is true then dnt will be false,
+        else by default dnt value is true
+      */
+      payload.dnt = !payload.att_authorization_status;
+    } else {
+      throw new CustomError("[Singular] :: Invalid Platform", 400);
+    }
+
+    // Singular maps Connection Type to either wifi or carrier
+    if (message.context?.network?.wifi === "wifi") {
+      payload.c = "wifi";
+    } else {
+      payload.c = "carrier";
+    }
+
+    payload = removeUndefinedAndNullAndEmptyValues(payload);
+    endPoint = `${BASE_URL}/launch`;
+  } else {
+    // event payload
+    if (platform.toLowerCase() === "android") {
+      payload = constructPayload(
+        message,
+        MAPPING_CONFIG[CONFIG_CATEGORIES.EVENT_ANDROID.name]
+      );
+      if (!payload) {
+        // fail-safety for developer error
+        throw new CustomError("Failed to Create Android Event Payload", 400);
+      }
+      extractCustomFields(
+        message,
+        eventAttributes,
+        ["properties"],
+        SINGULAR_EVENT_ANDROID_EXCLUSION
+      );
+    } else if (platform.toLowerCase() === "ios") {
+      payload = constructPayload(
+        message,
+        MAPPING_CONFIG[CONFIG_CATEGORIES.EVENT_IOS.name]
+      );
+      if (!payload) {
+        // fail-safety for developer error
+        throw new CustomError("Failed to Create iOS Event Payload", 400);
+      }
+      extractCustomFields(
+        message,
+        eventAttributes,
+        ["properties"],
+        SINGULAR_EVENT_IOS_EXCLUSION
+      );
+    } else {
+      throw new CustomError("[Singular] :: Invalid Platform", 400);
+    }
+    const productList = message.properties?.products;
+    if (productList && Array.isArray(productList)) {
+      const responseArray = [];
+      const finalPayload = payload;
+      for (const product of productList) {
+        const productDetails = constructPayload(
+          product,
+          MAPPING_CONFIG[CONFIG_CATEGORIES.PRODUCT_PROPERTY.name]
+        );
+        payload = { ...payload, ...productDetails };
+        const response = defaultRequestConfig();
+        response.endpoint = `${BASE_URL}/evt`;
+        payload = removeUndefinedAndNullAndEmptyValues(payload);
+        response.params = { ...payload, a: Config.apiKey, e: eventAttributes };
+        response.method = defaultGetRequestConfig.requestMethod;
+        responseArray.push(response);
+        payload = finalPayload;
```

Comment:
I think we can remove this line and directly put `finalPayload` above.

Reply:
these have been done to avoid extra properties from the previous products, coming into the new product.

Reply:
For each of the products, we are creating `productDetails` so how properties of the previous product will be there?

Reply:
> "product_id": "123",
        "products": [
            {
                "product_id": "789",
                "sku": "G-32",
                "name": "Monopoly",
                "price": 14,
                "quantity": 1,
                "category": "Games",
                "url": "https://www.website.com/product/path",
                "image_url": "https://www.website.com/product/path.jpg"
            },
            {
                "sku": "F-32",
                "name": "UNO",
                "price": 3.45,
                "quantity": 2,
                "category": "Games"
            }
        ]
        
These were done to handle this case: 
the second product doesn't have the product_id so we're taking product_id from the root.